File: ios/src/model/CameraInfo.swift
//
// © 2026-present https://github.com/cengiz-pz
//

import AVFoundation
import Foundation

@objc public final class CameraInfo: NSObject {
	@objc let id: String
	@objc let device: AVCaptureDevice
	@objc let outputSizes: [FrameSize]

	/// The clockwise angle in degrees (0, 90, 180, or 270) that the camera
	/// sensor image must be rotated to be upright when the device is held in
	/// its natural (portrait) orientation.
	///
	/// On iPhone, all built-in cameras capture natively in landscape, so this
	/// value is 90°.  On some iPad models the camera may be mounted differently
	/// and return 0°.  The value is a fixed hardware property and does not
	/// change with device rotation.  It is used alongside `isFrontFacing` to
	/// implement custom frame-processing pipelines that must account for sensor
	/// mounting angle independently of live device orientation.
	@objc let sensorOrientation: Int

	@objc(initWithId:device:outputSizes:sensorOrientation:)
	init(id cameraId: String, device: AVCaptureDevice, outputSizes: [FrameSize], sensorOrientation: Int) {
		self.id = cameraId
		self.device = device
		self.outputSizes = outputSizes
		self.sensorOrientation = sensorOrientation
		super.init()
	}
}

File: android/src/main/java/org/godotengine/plugin/nativecamera/model/CameraInfo.java
//
// © 2026-present https://github.com/cengiz-pz
//

package org.godotengine.plugin.nativecamera.model;

import android.util.Size;

import android.graphics.ImageFormat;
import android.hardware.camera2.CameraCharacteristics;

import java.util.ArrayList;
import java.util.List;

import org.godotengine.godot.Dictionary;


public class CameraInfo {
	private static final String CLASS_NAME = CameraInfo.class.getSimpleName();
	private static final String LOG_TAG = "godot::" + CLASS_NAME;

	private static final String DATA_CAMERA_ID_PROPERTY = "camera_id";
	private static final String DATA_IS_FRONT_FACING_PROPERTY = "is_front_facing";
	private static final String DATA_OUTPUT_SIZES_PROPERTY = "output_sizes";
	private static final String DATA_SENSOR_ORIENTATION_PROPERTY = "sensor_orientation";

	private static final String DATA_ZOOM_RATIO_RANGE_PROPERTY = "zoom_ratio_range";
	private static final String DATA_AVAILABLE_ZOOM_RATIOS_PROPERTY = "available_zoom_ratios"; // опционально

	private String cameraId;
	private CameraCharacteristics characteristics;

	public CameraInfo(String cameraId, CameraCharacteristics characteristics) {
		this.cameraId = cameraId;
		this.characteristics = characteristics;
	}

	public Dictionary buildRawData() {
		Dictionary dict = new Dictionary();

		dict.put(DATA_CAMERA_ID_PROPERTY, cameraId);

		Integer facing = characteristics.get(CameraCharacteristics.LENS_FACING);
		dict.put(DATA_IS_FRONT_FACING_PROPERTY,
				facing != null && facing == CameraCharacteristics.LENS_FACING_FRONT);

		android.hardware.camera2.params.StreamConfigurationMap map =
				characteristics.get(CameraCharacteristics.SCALER_STREAM_CONFIGURATION_MAP);

		List<Dictionary> dictList = new ArrayList<>();

		if (map != null) {
			Size[] sizes = map.getOutputSizes(ImageFormat.YUV_420_888);
			if (sizes != null) {
				for (Size size : sizes) {
					dictList.add(new FrameSize(size).buildRawData());
				}
			}
		}

		dict.put(DATA_OUTPUT_SIZES_PROPERTY, dictList.toArray());

		// SENSOR_ORIENTATION is the clockwise angle (0, 90, 180, or 270 degrees) that
		// the camera image must be rotated to be upright when the device is held in its
		// natural (portrait) orientation.  Null-safe: defaults to 0 when the value is
		// unavailable (e.g. on emulators or in unit tests).
		Integer sensorOrientation = characteristics.get(CameraCharacteristics.SENSOR_ORIENTATION);
		dict.put(DATA_SENSOR_ORIENTATION_PROPERTY, sensorOrientation != null ? sensorOrientation : 0);

		// Основной диапазон зума (самое важное)
		float[] zoomRatioRange = characteristics.get(CameraCharacteristics.CONTROL_ZOOM_RATIO_RANGE);
		if (zoomRatioRange != null && zoomRatioRange.length == 2) {
			Dictionary zoomDict = new Dictionary();
			zoomDict.put("min", zoomRatioRange[0]);
			zoomDict.put("max", zoomRatioRange[1]);
			dict.put(DATA_ZOOM_RATIO_RANGE_PROPERTY, zoomDict);
		}

		// Если камера поддерживает дискретные значения (редко, но бывает)
		if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
			float[] availableZooms = characteristics.get(CameraCharacteristics.CONTROL_AVAILABLE_ZOOM_RATIOS);
			if (availableZooms != null) {
				dict.put(DATA_AVAILABLE_ZOOM_RATIOS_PROPERTY, availableZooms);
			}
		}

		return dict;
	}
}
